TokenBucket.php

```diff
@@ -0,0 +1,140 @@
+<?
+
+namespace iFixit\TokenBucket;
+
+use \InvalidArgumentException;
+
+require_once __DIR__ . DIRECTORY_SEPARATOR . 'TokenRate.php';
+require_once __DIR__ . DIRECTORY_SEPARATOR . 'Backend.php';
+
+/**
+ * Implements the token bucket algorithm and stores tokens in storage.
+ *
+ * The rate of token regeneration is specified by the TokenRate class. This
+ * also specifies the maximum number of tokens the bucket can hold.
+ *
+ * The current number of tokens is time sensative so it is not stored as a
+ * member of the class, instead you will have to call getTokens() to calculate
+ * and retrieve the current number of tokens.
+ *
+ * Last consume can possibly be changed by another concurrent request, and
+ * therefore is also not stored in the Object's state and always retrieved from
+ * storage when needed.
+ */
+class TokenBucket {
+   // String to identify this bucket uniquely in storage.
+   private $key;
+   // TokenRate is when your parents try to ration your arcade spending.
+   private $rate;
+   // Where the token is to be stored.
+   private $backend;
+
+   /**
+    * @param $identifier string to uniqely store the bucket.
+    */
+   public function __construct($key, Backend $backend, TokenRate $rate) {
+      if (!is_string($key)) {
+         throw new InvalidArgumentException("identifier must be a string");
+      }
+
+      $this->key = $key;
+      $this->backend = $backend;
+      $this->rate = $rate;
+   }
+
+   /**
+    * Tries to consume a token, if a token can't be consumed, then a DateTime
+    * is returned for when the next token will be available.
+    *
+    * @return array index 0 being a boolean whether or not a token was consumed
+    *               index 1 being a timestamp of when there will be
+    *               enough tokens to consume the amount requested.
+    */
+   public function consume($amount) {
+      if (!is_int($amount)) {
+         throw new InvalidArgumentException("amount must be an int");
+      }
+
+      $now = microtime(true);
+      $storedBucket = $this->backend->get($this->key);
+      if ($storedBucket === Backend::MISS) {
+         $storedBucket = new StoredBucket($this->rate->getTokens(), $now);
+      }
+      $tokens = $storedBucket->getTokens();
+      $lastConsume = $storedBucket->getLastConsume();
+
+      $updatedTokens = $tokens - $amount;
+      $newBucket = new StoredBucket($updatedTokens, $now);
```

Comment:
@djmetzle pointed out that this will cause a bug in readyTime if we pass in a negative $updatedTokens. This should never go below zero. Some good thinking on his part. Recommended code should `$newBucket = new StoredBucket(min($updatedTokens, 0), $now);`